packages/apollo-forest-run/src/__tests__/eviction.test.ts

```diff
@@ -0,0 +1,145 @@
+  const result1 = cache.read({ query, variables: { i: 1 }, optimistic: true });
+  const result2 = cache.read({ query, variables: { i: 2 }, optimistic: true });
+
+  expect(result0).toEqual(null);
+  expect(result1).toEqual(null);
+  expect(result2).toEqual({ foo: 2 });
+});
+
+it("allows disabling automatic eviction", () => {
+  const cache = new ForestRun({
+    maxOperationCount: 1,
+    autoEvict: false,
+  });
+  const query = gql`
+    query ($i: Int) {
+      foo(i: $i)
+    }
+  `;
+
+  cache.write({ query, variables: { i: 0 }, result: { foo: 0 } });
+  cache.write({ query, variables: { i: 1 }, result: { foo: 1 } });
+  cache.write({ query, variables: { i: 2 }, result: { foo: 2 } });
+
+  const result0 = cache.read({ query, variables: { i: 0 }, optimistic: true });
+  const result1 = cache.read({ query, variables: { i: 1 }, optimistic: true });
+  const result2 = cache.read({ query, variables: { i: 2 }, optimistic: true });
+
+  expect(result0).toEqual({ foo: 0 });
+  expect(result1).toEqual({ foo: 1 });
+  expect(result2).toEqual({ foo: 2 });
+});
+
+it("allows manual eviction", () => {
+  const cache = new ForestRun({
+    maxOperationCount: 1,
+    autoEvict: false,
+  });
+  const query = gql`
+    query ($i: Int) {
+      foo(i: $i)
+    }
+  `;
+
+  cache.write({ query, variables: { i: 0 }, result: { foo: 0 } });
+  cache.write({ query, variables: { i: 1 }, result: { foo: 1 } });
+  cache.write({ query, variables: { i: 2 }, result: { foo: 2 } });
+
+  cache.gc();
+
+  const result0 = cache.read({ query, variables: { i: 0 }, optimistic: true });
+  const result1 = cache.read({ query, variables: { i: 1 }, optimistic: true });
+  const result2 = cache.read({ query, variables: { i: 2 }, optimistic: true });
+
+  expect(result0).toEqual(null);
+  expect(result1).toEqual(null);
+  expect(result2).toEqual({ foo: 2 });
+});
+
+it("doesn't evict watched old operations", () => {
+  const cache = new ForestRun({
+    maxOperationCount: 1,
+    autoEvict: false,
+  });
+  const query = gql`
+    query ($i: Int) {
+      foo(i: $i)
+    }
+  `;
+
+  cache.watch({
+    query,
+    variables: { i: 1 },
+    optimistic: true,
+    callback: () => {},
+  });
+
+  cache.write({ query, variables: { i: 0 }, result: { foo: 0 } });
+  cache.write({ query, variables: { i: 1 }, result: { foo: 1 } });
+  cache.write({ query, variables: { i: 2 }, result: { foo: 2 } });
+
+  cache.gc();
+
+  const result0 = cache.read({ query, variables: { i: 0 }, optimistic: true });
+  const result1 = cache.read({ query, variables: { i: 1 }, optimistic: true });
+  const result2 = cache.read({ query, variables: { i: 2 }, optimistic: true });
+
+  expect(result0).toEqual(null);
+  expect(result1).toEqual({ foo: 1 });
+  expect(result2).toEqual({ foo: 2 });
+});
+
+it("allows to opt out operations from eviction", () => {
+  const cache = new ForestRun({
+    maxOperationCount: 1,
+    autoEvict: false,
+    nonEvictableQueries: new Set(["b"]), // root-level field
```

Comment:
question: maxOperationCount is actually increased by length of nonEvictableQueries, right?

Reply:
yes, by design